### PR TITLE
 JSONSchema Fetching From Live Jenkins Instance.

### PR DESCRIPTION
**Pull request by sladyn98**
This PR deals with the fetching of the Schema from a live Jenkins instance.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,7 +10,8 @@
 		"Other"
 	],
 	"activationEvents": [
-		"onCommand:extension.helloWorld"
+		"onCommand:extension.helloWorld",
+		"onLanguage:json"
```
Comment (timja) [inline package.json]: some more background on what this does?

Comment (sladyn98) [reply package.json]: This activates the plugin on the command hello world, and on the language json, IMO it should be yaml and for the command we should keep it something simple.

Comment (timja) [reply package.json]: Yes, yaml and whatever commands we want to add...

File: src/extension.ts [hunk 2]
```diff
@@ -13,11 +13,11 @@ export function activate(context: vscode.ExtensionContext) {
 	// The command has been defined in the package.json file
 	// Now provide the implementation of the command with registerCommand
 	// The commandId parameter must match the command field in package.json
-	let disposable = vscode.commands.registerCommand('extension.helloWorld', () => {
+	let disposable = vscode.commands.registerCommand('extension.jcasc', () => {
 		// The code you place here will be executed every time your command is executed
```
Comment (sladyn98) [inline src/extension.ts]: We need to fetch the Schema here.Couple of questions
a) Do we need to store the schema anywhere?
b) Cache It.
@timja Do I need to run the jenkins instance or take the url as input ?

Comment (timja) [reply src/extension.ts]: a) yes in the vscode project folder i guess?
b) if someone asks for it again  they probably want a new one?

c) it should come from configuration, the user won't want to input  it every time

File: src/extension.ts [hunk 3]
```diff
@@ -18,6 +21,11 @@ export function activate(context: vscode.ExtensionContext) {
 
 		// Display a message box to the user
 		vscode.window.showInformationMessage('Welcome to the jcasc-plugin');
+
+		//Fetches the JSON Schema via a REST API Call
+		const endpoint = 'http://localhost:8080/jenkins/configuration-as-code/schema';
+		let resonse = this.get(endpoint);
+		console.log(JSON.stringify(resonse));
```
Comment (sladyn98) [inline src/extension.ts]: @timja This is just for trial, I  ran the jenkins instance and tried fetching the api, but it crashes before reaching this statement, I tried debugging but it seems like it cannot even start the session.

Comment (timja) [reply src/extension.ts]: try post instead of get?

**Comment by sladyn98** [comment]
@timja Should we take the jenkins URL from the User by prompting him to enter it in the workspace settings ?

**Comment by sladyn98** [comment]
Also Should we store the schema in the current working directory ? Since it would be easier to get detected by vscode then ?

**Comment by timja** [comment]
> Also Should we store the schema in the current working directory ? Since it would be easier to get detected by vscode then ?

Yeah probably either current working directory or .vscode/ in the project

**Comment by timja** [comment]
> @timja Should we take the jenkins URL from the User by prompting him to enter it in the workspace settings ?

yes

**Comment by sladyn98** [comment]
@timja Saved file to cwd

File: package.json [hunk 4]
```diff
@@ -10,14 +10,26 @@
...
+		"onLanguage:yml"
 	],
 	"main": "./out/extension.js",
 	"contributes": {
-		"commands": [{
-			"command": "extension.helloWorld",
-			"title": "Hello World"
-		}]
+		"configuration":{
+			"title": "JCasC",
```
Comment (timja) [inline package.json]:
```suggestion
			"title": "Jenkins: Configuration as Code",
```

File: package.json [hunk 5]
```diff
@@ -10,14 +10,26 @@
...
+			"properties": {
+				"jcasc.schemaURL": {
+					"type":"string",
+					"description": "URL for the JSON Schema for the Jenkins instance"
```
Comment (timja) [inline package.json]:
```suggestion
					"description": "URL for the configuration-as-code Schema for the Jenkins instance"
```

File: package.json [hunk 6]
```diff
@@ -10,14 +10,26 @@
 		"Other"
 	],
 	"activationEvents": [
-		"onCommand:extension.helloWorld"
+		"onCommand:extension.jcasc",
+		"onLanguage:yml"
```
Comment (timja) [inline package.json]: Just remove for now? I don't think it's doing anything atm
```suggestion
```

File: package.json [hunk 7]
```diff
@@ -10,14 +10,26 @@
...
+				"jcasc.schemaURL": {
+					"type":"string",
+					"description": "URL for the JSON Schema for the Jenkins instance"
+				}
+			}
+		},
+		"commands": [
+			{
+				"command": "extension.jcasc",
+				"title": "JCasC"
```
Comment (timja) [inline package.json]:
```suggestion
				"title": "Jenkins: Fetch configuration-as-code schema"
```

File: src/extension.ts [hunk 8]
```diff
@@ -13,11 +20,27 @@ export function activate(context: vscode.ExtensionContext) {
...
 
 		// Display a message box to the user
-		vscode.window.showInformationMessage('Hello World!');
+		vscode.window.showInformationMessage('Welcome to the jcasc-plugin');
+		const schemaURL = vscode.workspace.getConfiguration().get('jcasc.schemaURL');
+		//Fetches the JSON Schema via a REST API Call
+		rp(schemaURL).then((result:any) => {
+			console.log("Result: " + result);
+			var fileContent = result;
+			fs.writeFile("jsonSchema.json", fileContent, (err:any) => {
```
Comment (timja) [inline src/extension.ts]:
```suggestion
			fs.writeFile("jcasc-schema.json", fileContent, (err:any) => {
```

**Pull request merged**